File: hdd/reverse_string/answer.dart
class Solution {
  int reverse(int x) {

    // 숫자 범위
    if(x == 0){
      return 0;
    }

    if(x.abs() % 10 == 0){
      return divideByTen(x.abs() ~/ 10) * x.sign;
    }else {
      return divideByTen(x.abs()) * x.sign;
    }
  }

  divideByTen(int x){
    String cachedString = "";
    var temp = x;
    while(temp > 0){
      cachedString = "$cachedString${temp%10}";
      temp = temp ~/ 10;
    }


    var result = int.parse(cachedString);

    if(result <= pow(-2, 31) || result >= (pow(2,31) - 1)){
      return 0;
    }
    return result;
  }
}

//Runtime: 469 ms, faster than 100.00% of Dart online submissions for Reverse Integer.
//Memory Usage: 143.2 MB, less than 100.00% of Dart online submissions for Reverse Integer.

File: hdd/reverse_string/answer.kt
import kotlin.math.sign
import kotlin.math.abs

class Solution {
    fun reverse(x: Int): Int {
        if(x == 0){
            0
        }

        val abs = abs(x)
        if(abs % 10 == 0){
            return divideByTen(abs / 10) * x.sign;
        }else {
            return divideByTen(abs) * x.sign;
        }
    }

    fun divideByTen(x: Int) : Int{
        var cachedString = ""
        var temp = x;

        while(temp > 0){
            cachedString = "$cachedString${temp%10}"
            temp = temp / 10
        }

        return cachedString.toIntOrNull() ?: 0
    }
}

//Runtime: 187 ms, faster than 84.14% of Kotlin online submissions for Reverse Integer.
//Memory Usage: 35.6 MB, less than 30.62% of Kotlin online submissions for Reverse Integer.